File: codes/semana12/PokemonBattle_factory.py
from __future__ import annotations
from abc import ABC, abstractmethod
from dataclasses import dataclass, field


# ── Produto base ──────────────────────────────────────────
@dataclass
class Pokemon(ABC):
    name:   str
    hp:     int
    attack: int

    @abstractmethod
    def use_move(self) -> None: ...

    def status(self) -> str:
        return (f"{self.name:<14}"
                f"HP:{self.hp:>4}  "
                f"ATK:{self.attack:>4}")


# ── Produtos concretos ────────────────────────────────────
class FirePokemon(Pokemon):
    def use_move(self) -> None:
        print(f"{self.name} usou LANÇA-CHAMAS! "
              f"(dano base: {self.attack})")

class WaterPokemon(Pokemon):
    def use_move(self) -> None:
        print(f"{self.name} usou HIDROBOMBA! "
              f"(dano base: {self.attack})")

class GrassPokemon(Pokemon):
    def use_move(self) -> None:
        print(f"{self.name} usou FOLHA-NAVALHA! "
              f"(dano base: {self.attack})")

class PsychicPokemon(Pokemon):
    def use_move(self) -> None:
        print(f"{self.name} usou PSIQUICO! "
              f"(dano base: {self.attack})")


# ── Simple Factory com registry ───────────────────────────
class PokemonFactory:
    _registry: dict[str, type[Pokemon]] = {
        "fire":    FirePokemon,
        "water":   WaterPokemon,
        "grass":   GrassPokemon,
        "psychic": PsychicPokemon,
    }

    @classmethod
    def catch(
        cls,
        poke_type: str,
        name: str,
        hp: int,
        attack: int,
    ) -> Pokemon:
        if poke_type not in cls._registry:
            raise ValueError(
                f"Tipo {poke_type!r} nao existe na Pokedex!")
        return cls._registry[poke_type](
            name=name, hp=hp, attack=attack)

    @classmethod
    def register(
        cls,
        poke_type: str,
        klass: type[Pokemon],
    ) -> None:
        """Adiciona novo tipo sem alterar codigo existente."""
        cls._registry[poke_type] = klass


# ── Uso ───────────────────────────────────────────────────
if __name__ == "__main__":
    team_data = [
        ("fire",    "Charizard",  78, 84),
        ("water",   "Blastoise",  79, 83),
        ("grass",   "Venusaur",   80, 82),
        ("psychic", "Mewtwo",    106, 110),
    ]

    print("=== Status do Time ===")
    team = [PokemonFactory.catch(*d) for d in team_data]
    for p in team:
        print(p.status())

    print("=== Batalha! ===")
    for p in team:
        p.use_move()

    # Extensao dinamica — novo tipo sem alterar a factory
    class DragonPokemon(Pokemon):
        def use_move(self) -> None:
            print(f"{self.name} usou PULSO-DRAGAO! "
                  f"(dano base: {self.attack})")

    PokemonFactory.register("dragon", DragonPokemon)
    dragonite = PokemonFactory.catch(
        "dragon", "Dragonite", hp=91, attack=134)

    print("=== Novo tipo registrado ===")
    print(dragonite.status())
    dragonite.use_move()

    # Tipo invalido
    try:
        PokemonFactory.catch("electric", "Pikachu", 35, 55)
    except ValueError as e:
        print(f"Erro: {e}")

File: codes/semana12/PokemonBattle_factory.cpp
#include <iostream>
#include <memory>
#include <string>
#include <map>
#include <functional>
#include <stdexcept>
#include <iomanip>
#include <sstream>
#include <vector>

// ── Produto base ──────────────────────────────────────────
// @dataclass ABC  →  classe com construtor explícito
//                    e método puramente virtual (= 0)
class Pokemon {
protected:
    std::string name_;
    int         hp_;
    int         attack_;

public:
    Pokemon(const std::string& name, int hp, int attack)
        : name_(name), hp_(hp), attack_(attack) {}

    // @abstractmethod use_move  →  virtual puro
    virtual void use_move() const = 0;

    // f"{self.name:<14}HP:{self.hp:>4}  ATK:{self.attack:>4}"
    // →  std::setw + std::left / std::right (iomanip)
    std::string status() const {
        std::ostringstream oss;
        oss << std::left  << std::setw(14) << name_
            << "HP:"
            << std::right << std::setw(4)  << hp_
            << "  ATK:"
            << std::right << std::setw(4)  << attack_;
        return oss.str();
    }

    virtual ~Pokemon() = default;
};

// ── Produtos concretos ────────────────────────────────────
class FirePokemon : public Pokemon {
public:
    using Pokemon::Pokemon;   // herda o construtor
    void use_move() const override {
        std::cout << name_ << " usou LANÇA-CHAMAS! "
                  << "(dano base: " << attack_ << ")";
    }
};

class WaterPokemon : public Pokemon {
public:
    using Pokemon::Pokemon;
    void use_move() const override {
        std::cout << name_ << " usou HIDROBOMBA! "
                  << "(dano base: " << attack_ << ")";
    }
};

class GrassPokemon : public Pokemon {
public:
    using Pokemon::Pokemon;
    void use_move() const override {
        std::cout << name_ << " usou FOLHA-NAVALHA! "
                  << "(dano base: " << attack_ << ")";
    }
};

class PsychicPokemon : public Pokemon {
public:
    using Pokemon::Pokemon;
    void use_move() const override {
        std::cout << name_ << " usou PSÍQUICO! "
                  << "(dano base: " << attack_ << ")";
    }
};

// ── Simple Factory com registry ───────────────────────────
// dict[str, type[Pokemon]]
// →  map<string, function<unique_ptr<Pokemon>(string,int,int)>>
//
// Em Python a classe é armazenada e chamada diretamente.
// Em C++ armazenamos uma lambda criadora — mesmo efeito.
class PokemonFactory {
    // Tipo do criador: recebe (name, hp, attack) → unique_ptr
    using Creator = std::function<
        std::unique_ptr<Pokemon>(
            const std::string&, int, int)>;

    // registry estático local (equivale ao atributo de classe)
    static std::map<std::string, Creator>& registry() {
        static std::map<std::string, Creator> reg = {
            {"fire", [](const std::string& n, int hp, int atk){
                return std::make_unique<FirePokemon>(n, hp, atk);
            }},
            {"water", [](const std::string& n, int hp, int atk){
                return std::make_unique<WaterPokemon>(n, hp, atk);
            }},
            {"grass", [](const std::string& n, int hp, int atk){
                return std::make_unique<GrassPokemon>(n, hp, atk);
            }},
            {"psychic", [](const std::string& n, int hp, int atk){
                return std::make_unique<PsychicPokemon>(n, hp, atk);
            }},
        };
        return reg;
    }

public:
    // @classmethod catch  →  static catch_
    // (catch é palavra reservada em C++, usamos catch_)
    static std::unique_ptr<Pokemon>
    catch_(const std::string& poke_type,
           const std::string& name,
           int hp, int attack)
    {
        auto it = registry().find(poke_type);
        if (it == registry().end())
            // raise ValueError  →  throw invalid_argument
            throw std::invalid_argument(
                "Tipo '" + poke_type
                + "' nao existe na Pokedex!");
        return it->second(name, hp, attack);
    }

    // @classmethod register  →  static registerType
    static void registerType(
        const std::string& poke_type, Creator fn)
    {
        registry()[poke_type] = fn;
    }
};

// ── Uso ───────────────────────────────────────────────────
int main() {
    // team_data = [("fire","Charizard",78,84), ...]
    using TeamEntry = std::tuple<
        std::string, std::string, int, int>;

    std::vector<TeamEntry> team_data = {
        {"fire",    "Charizard",  78,  84},
        {"water",   "Blastoise",  79,  83},
        {"grass",   "Venusaur",   80,  82},
        {"psychic", "Mewtwo",    106, 110},
    };

    std::cout << "=== Status do Time ===";
    std::vector<std::unique_ptr<Pokemon>> team;
    for (const auto& [type, name, hp, atk] : team_data) {
        auto p = PokemonFactory::catch_(type, name, hp, atk);
        std::cout << p->status() << "";
        team.push_back(std::move(p));
    }

    std::cout << "=== Batalha! ===";
    for (const auto& p : team)
        p->use_move();

    // Extensão dinâmica — classe local dentro de main()
    // equivale à subclasse inline do Python
    class DragonPokemon : public Pokemon {
    public:
        using Pokemon::Pokemon;
        void use_move() const override {
            std::cout << name_ << " usou PULSO-DRAGÃO! "
                      << "(dano base: " << attack_ << ")";
        }
    };

    PokemonFactory::registerType("dragon",
        [](const std::string& n, int hp, int atk){
            return std::make_unique<DragonPokemon>(n, hp, atk);
        });

    auto dragonite = PokemonFactory::catch_(
        "dragon", "Dragonite", 91, 134);

    std::cout << "=== Novo tipo registrado ===";
    std::cout << dragonite->status() << "";
    dragonite->use_move();

    // try / except ValueError  →  try / catch invalid_argument
    try {
        PokemonFactory::catch_("electric", "Pikachu", 35, 55);
    } catch (const std::invalid_argument& e) {
        std::cout << "Erro: " << e.what() << "";
    }

    return 0;
}